File: practice/implementation/matrix_rotation/matrix_rotation.py
# Enter your code here. Read input from STDIN. Print output to STDOUT
def transfrom(i_matrix, i_m, i_n):
    res = []
    num_rows = i_m
    num_cols = i_n
    min_row = 0
    max_row = num_rows - 1
    min_col = 0
    max_col = num_cols - 1
    while min_row <= max_row and min_col <= max_col:
        tmp = []
        for ROW_NUM in range(min_row, max_row):
            tmp.append(i_matrix[ROW_NUM][min_col])
        for COL_NUM in range(min_col, max_col):
            tmp.append(i_matrix[max_row][COL_NUM])
        for ROW_NUM in range(max_row, min_row, -1):
            tmp.append(i_matrix[ROW_NUM][max_col])
        for COL_NUM in range(max_col, min_col, -1):
            tmp.append(i_matrix[min_row][COL_NUM])
        res.append([VAL for VAL in tmp])
        min_row += 1
        max_row -= 1
        min_col += 1
        max_col -= 1
    return res

def inverse_transform(i_trans_matrix, i_m, i_n):
    res = [[0 for _ in range(i_n)] for __ in range(i_m)]
    num_rows = i_m
    num_cols = i_n
    min_row = 0
    max_row = num_rows - 1
    min_col = 0
    max_col = num_cols - 1
    for LINE in i_trans_matrix:
        cnt = 0
        for ROW_NUM in range(min_row, max_row):
            res[ROW_NUM][min_col] = LINE[cnt]
            cnt += 1
        for COL_NUM in range(min_col, max_col):
            res[max_row][COL_NUM] = LINE[cnt]
            cnt += 1
        for ROW_NUM in range(max_row, min_row, -1):
            res[ROW_NUM][max_col] = LINE[cnt]
            cnt += 1
        for COL_NUM in range(max_col, min_col, -1):
            res[min_row][COL_NUM] = LINE[cnt]
            cnt += 1
        min_row += 1
        max_row -= 1
        min_col += 1
        max_col -= 1
    return res

def rotate(i_trans_matrix, i_r):
    rotated = []
    for CNT in range(len(i_trans_matrix)):
        L = len(i_trans_matrix[CNT])
        r = i_r % L
        #print i_trans_matrix[CNT]
        rotated.append([VAL for VAL in i_trans_matrix[CNT][L - r:] + i_trans_matrix[CNT][:L - r]])
        #print rotated[-1]
        #assert L == len(rotated[-1])
    return rotated

m, n, r = map(int, raw_input().split())
matrix = []
for _ in range(m):
    matrix.append(map(int, raw_input().split()))
    
t = transfrom(matrix, m, n)
rot = rotate(t, r)
res = inverse_transform(rot, m , n)
for LINE in res:
    for VAL in LINE:
        print VAL,
    print

File: practice/algorithms/implementation/matrix_rotation/matrix_rotation.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;

vector<vector<int>> transfrom(const vector<vector<int>> & matrix, int num_rows, int num_cols)
{
    vector<vector<int>> res;
    int min_row = 0;
    int max_row = num_rows - 1;
    int min_col = 0;
    int max_col = num_cols - 1;
        
    while (min_row <= max_row && min_col <= max_col)
    {
        vector<int> tmp_row;
        for (int rn = min_row; rn < max_row; rn++)
        {
            tmp_row.push_back(matrix[rn][min_col]);
        }
        for (int cn = min_col; cn < max_col; cn++)
        {
            tmp_row.push_back(matrix[max_row][cn]);
        }
        for (int rn = max_row; rn > min_row; rn--)
        {
            tmp_row.push_back(matrix[rn][max_col]);
        }
        for (int cn = max_col; cn > min_col; cn--)
        {
            tmp_row.push_back(matrix[min_row][cn]);
        }
        res.push_back(tmp_row);
        
        min_row++;
        max_row--;
        min_col++;
        max_col--;
    }
        
    return res;
}
        
vector<vector<int>> inv_transfrom(const vector<vector<int>> & trans_matrix, int num_rows, int num_cols)
{
    vector<vector<int>> res(num_rows, vector<int>(num_cols));
    int min_row = 0;
    int max_row = num_rows - 1;
    int min_col = 0;
    int max_col = num_cols - 1;
    
    for (const auto & row : trans_matrix)
    {
        int cnt = 0;
        for (int rn = min_row; rn < max_row; rn++)
        {
            res[rn][min_col] = row[cnt];
            cnt++;
        }
        for (int cn = min_col; cn < max_col; cn++)
        {
            res[max_row][cn] = row[cnt];
            cnt++;
        }
        for (int rn = max_row; rn > min_row; rn--)
        {
            res[rn][max_col] = row[cnt];
            cnt++;
        }
        for (int cn = max_col; cn > min_col; cn--)
        {
            res[min_row][cn] = row[cnt];
            cnt++;
        }
        
        min_row++;
        max_row--;
        min_col++;
        max_col--;
    }
     
    return res;
}

vector<vector<int>> rotate(const vector<vector<int>> & trans_matrix, unsigned r)
{
    vector<vector<int>> res;
    for (const auto & row : trans_matrix)
    {
        auto crow = row;
        const size_t m = row.size();
        const size_t d = r % m;
        rotate(crow.begin(), crow.begin() + (m - d), crow.end());
        res.push_back(crow);
    }
    return res;
}

int main()
{
    unsigned m, n, r;
    cin >> m >> n >> r;
    
    vector<vector<int>> matrix(m, vector<int>(n));
    for (int row = 0; row < m; row++)
    {
        for (int col = 0; col < n; col++)
        {
            cin >> matrix[row][col];
        }
    }
    
    auto tmatrix = transfrom(matrix, m, n);
    auto rotated = rotate(tmatrix, r);
    auto res = inv_transfrom(rotated, m, n);
    
    for (const auto & row : res)
    {
        for (const auto & val : row)
        {
            cout << val << " ";
        }
        cout << endl;
    }
    
    return 0;
}

